### PR TITLE
feat(vscode_projects): add vscode project plugin

### PR DESCRIPTION
**Pull request by cathaysia**
Signed-off-by: Longtao Zhang <DragonBillow@outlook.com>

### CODE REVIEW

File: vscode_projects/__init__.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
+"""
+This extension provides a quick introduction on how to use the new Python pluging interface.
```
Comment (ManuelSchneid3r) [inline vscode_projects/__init__.py:2]: this string appears in the settings and should serve as longer description of the plugin

File: vscode_projects/__init__.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+"""
+
+from albert import *
+import os
+
+import json
+import urllib.parse
+
+md_iid = "0.5"
+md_version = "1.1"
```
Comment (ManuelSchneid3r) [inline vscode_projects/__init__.py:13]: has there been a 1.0?

File: vscode_projects/__init__.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+md_name = "vscode projects"
+md_description = "vscode projects"
```
Comment (ManuelSchneid3r) [inline vscode_projects/__init__.py:15]: This is a human readable pretty printed string. i should be brief and imperative like "Open VSCode projects"

File: vscode_projects/__init__.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+from albert import *
+import os
+
+import json
+import urllib.parse
+
+md_iid = "0.5"
+md_version = "1.1"
+md_name = "vscode projects"
```
Comment (ManuelSchneid3r) [inline vscode_projects/__init__.py:14]: This is a human readable pretty printed string. Use VSCode projects

File: vscode_projects/__init__.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+md_description = "vscode projects"
+md_license = "BSD-3"
+md_url = "https://github.com/albertlauncher/python/tree/master/kill"
```
Comment (ManuelSchneid3r) [inline vscode_projects/__init__.py:17]: no not kill

File: vscode_projects/__init__.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+md_maintainers = "@cathaysia"
+md_credits = "Original idea by Longtao Zhang"
+
+vsc_db = "~/.config/Code/User/globalStorage/state.vscdb"
```
Comment (ManuelSchneid3r) [inline vscode_projects/__init__.py:21]: try to generalize this. this will work on linux only.

File: vscode_projects/__init__.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+vsc_storage = "~/.config/Code/User/globalStorage/storage.json"
+vsc_workspace = "~/.config/Code/User/workspaceStorage/"
+
+
+class Plugin(QueryHandler):
+    icon_path = "xdg:code"
+
+    def id(self):
+        return __name__
```
Comment (ManuelSchneid3r) [inline vscode_projects/__init__.py:30]: use md_id here

File: vscode_projects/__init__.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+    def name(self):
+        return md_name
+
+    def description(self):
+        return md_description
+
+    def initialize(self):
```
Comment (ManuelSchneid3r) [inline vscode_projects/__init__.py:38]: drop if unused

File: vscode_projects/__init__.py [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+                        continue
+                    results.append(
+                        Item(
+                            id="kill",
+                            icon=[self.icon_path],
+                            text=os.path.basename(full_prj_path),
+                            subtext=full_prj_path,
+                            actions=[
+                                Action(
+                                    "打开",
```
Comment (ManuelSchneid3r) [inline vscode_projects/__init__.py:68]: ?! this should be usable for the entire worldwide community

File: vscode_projects/__init__.py [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+                                    "使用 vscode 打开项目",
```
Comment (ManuelSchneid3r) [inline vscode_projects/__init__.py:69]: same here

**Review by ManuelSchneid3r — CHANGES_REQUESTED** [review]
This plugin is far from polished. Being in this unfinished state, why do you send this PR at all? Do you really expect me to merge and ship it? Please polish, otherwise i cant merge it.

**Comment by cathaysia** [comment]
OK, I give some changes on codes

File: vscode_projects/__init__.py [hunk 11]
```diff
@@ -17,18 +18,13 @@
...
 
 class Plugin(QueryHandler):
     icon_path = "xdg:code"
+    os_prefix = "~" if platform.system().lower() != "window" else os.getenv("APPDATA") or ""
+    db = os.path.expanduser(os_prefix + "/.config/Code/User/globalStorage/state.vscdb")
+    storage = os.path.expanduser(os_prefix + "/.config/Code/User/globalStorage/storage.json")
+    workspace = os.path.expanduser(os_prefix + "/.config/Code/User/workspaceStorage/")
 
     def id(self):
         return md_iid
```
Comment (ManuelSchneid3r) [inline vscode_projects/__init__.py:30]: iid is the interface identifier. Use md_id

**Review by ManuelSchneid3r — CHANGES_REQUESTED** [review]

File: vscode_projects/__init__.py [hunk 12]
```diff
@@ -61,7 +61,7 @@ def handleQuery(self, query):
                 uri = urllib.parse.unquote(uri)
                 results.append(
                     Item(
-                        id="kill",
+                        id=md_id,
```
Comment (ManuelSchneid3r) [inline vscode_projects/__init__.py:64]: this should be the identifier of a project item right? identifiers have to be unique.

**Review by ManuelSchneid3r — COMMENTED** [review]
I think you got that wrong. The identifier has to be unique such that each item has its own idententifier, use project ids or path or the name of the project (but actually this is probably not guaranteed to be unique as well)

**Comment by cathaysia** [comment]
> I think you got that wrong. The identifier has to be unique such that each item has its own idententifier, use project ids or path or the name of the project (but actually this is probably not guaranteed to be unique as well)

can I use uuid.uuid1() ?

**Comment by ManuelSchneid3r** [comment]
@cathaysia why did you close the pr? 

> can I use uuid.uuid1() ?

why not just sth like `'md_id' + project_path`?

The id has to be persistant (because it will be reused for sorting) and unique (because it is used as icon cache key).

**Comment by cathaysia** [comment]
I worry that I don't have much time to maintain this code. This plugin was originally written just because I needed this plugin. If someone is happy to maintain it, you can simply copy and modify my code.

**Comment by ManuelSchneid3r** [comment]
You dont necessarily have to maintain it. Its fine to leave the maintainer field empty, but initially the plugin should be polished before its shipped around the world.

**Pull request closed without merging**